RFC: Safety Tags

zjp-CN opened the pull request:
# Summary

This RFC introduces a concise safety-comment convention for unsafe code in standard libraries:
tag every public unsafe function with `#[safety::requires]` and call with `#[safety::checked]`.

Safety tags refine today’s safety-comment habits: a featherweight syntax that condenses every
requirement into a single, check-off reminder.

The following snippet [compiles] today if we enable enough nightly features, but we expect Clippy
and Rust-Analyzer to enforce tag checks and provide first-class IDE support.

[compiles]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=6b7b341b8879bfdecb80ae72a8011a6d

```rust
#[safety::requires( // 💡 define safety tags on an unsafe function
    valid_ptr = "src must be [valid](https://doc.rust-lang.org/std/ptr/index.html#safety) for reads",
    aligned = "src must be properly aligned, even if T has size 0",
    initialized = "src must point to a properly initialized value of type T"
)]
pub unsafe fn read<T>(ptr: *const T) { }

fn main() {
    #[safety::checked( // 💡 discharge safety tags on an unsafe call
        valid_ptr, aligned, initialized = "optional reason"
    )]
    unsafe { read(&()) };
}
```


[Rendered](https://github.com/safer-rust/rfcs/blob/safety-tags/text/0000-safety-tags.md)

zjp-CN committed:
RFC: safety-tags

zjp-CN committed:
safety-tags: styling

zjp-CN committed:
safety-tags: add previous zulipchat link on Safety Property System

zjp-CN committed:
safety-tags: add tag discharge of unsafe encapsulation

zjp-CN committed:
safety-tags: typo

zjp-CN committed:
safety-tags: styling

zjp-CN commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,471 @@
+```
+
+# Motivation
+[motivation]: #motivation
````
Thanks for comments! It's late in my timezone. I will try to update the document tomorrow.

zjp-CN replied:
I've updated the motivation section to include four problems on current safety practices.

juntyr commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,481 @@
+}
+
+#[clippy::safety { ValidPtr }] // 💡
+pub unsafe fn read<T>(ptr: *const T) {}
```
What about functions that take several parameters with some different, some overlapping requirements? Would it be possible to scope some tags to just some arguments (or maybe even type generics) to require the caller to discharge the requirements separately for each?

danjl1100 commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,481 @@
+   referenced tag is defined and accessible.
+3. Verify that every unsafe call carries the required safety tags:
+   * Resolve the callee, collect its declared tags, then walk outward from the call site until the
+     function’s own signature confirms these tags are listed in its `#![clippy::safety { ... }]`
```
On first read through, I was confused on the syntax examples separating imports (inner attribute `#![...]`) from function definitions and use sites (both outer attributes `#[...]`). 

Should this be an outer attribute `#[...]` since it's on a function signature? Or are both inner/outer attributes allowed on function definitions?

(I think this was the only place I noticed)

ChayimFriedman2 commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,481 @@
+Libraries in `rust-lang/rust` must enforce tag checking as a hard error, guaranteeing that every tag
+definition and discharge is strictly valid.
+
+## Implementation in Rust Analyzer
```
It feels weird to put the attributes under the `clippy` namespace if they are not a clippy-only thing. I can also expect other tools developed to analyze Rust will want to consider these attributes. I'd say put them under `diagnostics` or their own namespace.

zjp-CN replied:
I've replaced clippy namespace with safety namespace. Restriction of single-path tool attr is solved by separating safety `requires` and `checked`. Thanks to @kennytm for the [idea](https://github.com/rust-lang/rfcs/pull/3842#discussion_r2247361847).

zjp-CN commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,481 @@
+Before tagging a function, we must declare them as ordinary items with `#[clippy::safety::tag]` such
+as [uninhabited] types or plain functions:
+
+```rust
+#[clippy::safety::tag]
+enum ValidPtr {}
+
+#[clippy::safety::tag]
+fn Aligned() {}
+```
+
````
Good question!

> Should the tag be an uninhabited enum, or a function? Why does it just work on any kind of item?

Semantically, a tag is just an uninhabited enum—no values are ever constructed.  

For rustdoc, however, we want to render it as a function; the resulting page is cleaner and avoids the noise of auto-trait impls. [Link1](https://internals.rust-lang.org/t/pre-rfc-safety-property-system/23252/25) and [link2](https://internals.rust-lang.org/t/pre-rfc-safety-property-system/23252/27) to previous discussion on IRLO for rendering example.

> I still think it would be much clearer to have it as a macro call like `clippy::define_safety_tag!(Aligned);`

Therefore, we still need a (possibly built-in) proc-macro that ships with `libcore`; it hides every declaration detail and is the neat, single knob users turn.  

The alternative — treating tags as non-items and writing `#![clippy::safety::declare(Aligned)]`[^1] inside a module — would forfeit two things we want to reuse: the ordinary item namespace/visibility rules from `rustc` and existing item rendering from `rustdoc`.

[^1]: I would suggest this syntax for non-item-like tag declaration.

Lokathor replied:
If it's just a rustdoc issue, rustdoc can be smarter.

zjp-CN replied:
Thanks. I'll add contents on feature requests for rustdoc.

clarfonthey replied:
I should probably clarify what I meant by "a lot would need to happen," since I guess it's not a whole lot. Essentially, we would need "tool macros" to be a thing, similar to tool attributes, where the macros are explicitly ignored outside of the context of the tools.

This kind of tracks with how third-party "tools" work today: for example, `cargo-mutants` uses a no-op procedural attribute macro `mutants::skip` to mark things it should skip, and since the procedural macro is a no-op outside the tool, it doesn't affect anything. Theoretically, any tool could define a macro in a crate that does nothing and it would work fine here.

The biggest change is explicitly ignoring all macros namespaced under known tools like `clippy`, and then `clippy::define_safety_tag!` would work. Tools themselves can lint whether these macros are actually valid.

zjp-CN replied:
Oh, I see [`mutants::skip`](https://github.com/sourcefrog/cargo-mutants/blob/069006aeb2bc60f5d6cfa25ab3d307e43c57d7e1/mutants_attrs/src/lib.rs#L29-L31) and what you mean.

But I don't get why `mutants::skip` attribute needs to be defined as a proc-macro. Users of cargo-mutant can gate `mutant::skip` behind a conditional compilation:

```rust
#[cfg_attr(mutant, mutant::skip)]
fn foo() {}
```

And cargo-mutant passes extra `--cfg=mutant -Zcrate-attr=feature(register_tool) -Zcrate-attr=register_tool(mutant)` RUSTFLAGS when compiling the code. This is a usefull [trick](https://github.com/os-checker/distributed-verification/discussions/52) I learnt from kani for tool attributes.

---

While `clippy::define_safety_tag!` differs in that it produces ASTs, this is unlikely to occur: tool attributes never expand macros; expansion is the responsibility of procedural or declarative macros. (Or maybe of solely declarative macros in the future: [#3697](https://github.com/rust-lang/rfcs/pull/3697) and [#3698](https://github.com/rust-lang/rfcs/pull/3698))

clarfonthey replied:
The sole purpose of the proc macro is to avoid having to gate it behind a `cfg_attr` on stable (since `register_tool` is unstable), but you're correct that it's not strictly necessary. My main point was that since attributes can be macros, it's not too unreasonable to have tool macro calls in addition to tool *attribute* macro calls.

In the case of a macro call, the tool macros would not actually expand to anything regardless of whether they're used by a tool. It would mean that the implementation to the tool is different, but not extremely different: instead of collecting a list of items that have an attached attribute and their paths, you would instead collect a list of macro call items and then create the path from the combination of the call and the path of the module the call is made in.

zjp-CN replied:
I've updated the RFC with "[Unresolved questions - Encapsulate Tag Item Declaration with define_safety_tag!](https://github.com/Artisan-Lab/rfcs/blob/b077573c87122bb0972182d55a45648dca6853fd/text/0000-safety-tags.md#encapsulate-tag-item-declaration-with-define_safety_tag)" and "Future possibilities - [Better Rustdoc Rendering](https://github.com/Artisan-Lab/rfcs/blob/b077573c87122bb0972182d55a45648dca6853fd/text/0000-safety-tags.md#better-rustdoc-rendering)". Also worth mentioning that uninhabited enums are the only tag items now.

zjp-CN committed:
safety-tags: fix link

zjp-CN committed:
safety-tags: clarify semver rules

zjp-CN committed:
safety-tags: #[ref] needs clippy reporting

zjp-CN committed:
safety-tags: #![clippy::safety] => #![clippy::safety::use]

zjp-CN committed:
safety-tags: fix tag in loop snippet

zjp-CN committed:
safety-tags: clippy::safety::use => clippy::safety::r#use

error: expected identifier, found keyword `use`
 --> src/main.rs:2:20
  |
2 | #![clippy::safety::use(invariant::ValidPtr)] // 💡
  |                    ^^^ expected identifier, found keyword

zjp-CN committed:
safety-tags: clarify #[ref] support from clippy and RA

kennytm commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,485 @@
+We can also attach comments for a tag or a group of tags to clarify how safety requirements are met:
+
+```rust
+#[clippy::safety {
+  ValidPtr, Aligned, Initialized: "addr range p..p+n is properly initialized from aligned memory"
+  InBounded, ValidNum: "`n` won't exceed isize::MAX here, so `p.add(n)` is fine";
+}]
+for _ in 0..n {
+    unsafe {
````
The `#[clippy::safety]` attached to the `for` expression rather than the inner `unsafe`, is this expected? i'm not sure if allowing `#[clippy::safety]` be applied "at a distance" is a good idea

zjp-CN replied:
Yes. It's the expected behavior, as it's meaningful to explain how invariants are upheld in loops. [Previous discussion.](https://rust-lang.zulipchat.com/#narrow/channel/136281-t-opsem/topic/Safety.20Property.20System/near/530636749)

kennytm commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,485 @@
+```
+
+That's to say:
+* Tags declared or re-exported in the current module are automatically in scope: no import required.
````
how are tags reexported? 

does this mean if we `#[clippy::safety::r#use { core::ptr::invariants::* }] mod foo {}`, we can refer to them as `#[clippy::safety { foo::Aligned }] unsafe ...` ?

or does this mean if we `mod foo { pub use core::ptr::invariants::*; }`, we can refer to them as `#[clippy::safety { foo::Aligned }] unsafe ...` ?

kennytm commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+
+That's to say:
+* Tags declared or re-exported in the current module are automatically in scope: no import required.
+* Tags from other modules must be brought in with the inner-tool attribute shown above.
```
we can't `#[clippy::safety { core::ptr::invariants::Aligned }] unsafe ...` directly without using `r#use`?

zjp-CN replied:
We can restore `safety` without `r#use` to import tags if `safety::requires` and `safety::checked` are syntax for tagging.
[Previous discussion to why `r#use`.](https://github.com/rust-lang/rfcs/pull/3842#discussion_r2246032076)

zjp-CN replied:
Nit: `safety::r#use` won't happen on unsafe fns or expressions. So 

```diff
-#[clippy::safety { core::ptr::invariants::Aligned }]
+#[clippy::safety { core::ptr::invariants::Aligned }] unsafe ... 
```

zjp-CN replied:
But to be clearer, how about `#![clippy::safety::import(...)]` to avoid `use` keyword? For now, we have these syntax:

```rust
#[clippy::safety::import(...)]       # import tag items
#[clippy::safety::declare_tag(...)]  # declare a tag on items
#[clippy::safety::requires(...)]     # define tags on unsafe functions
#[clippy::safety::checked(...)]      # discharge tags on expressions containing unsafe calls
```

kennytm replied:
@zjp-CN my primary concern is that this item is saying

> <ins>① Tags from other modules</ins> must be brought in with the inner-tool <ins>② attribute shown above</ins>.

so it sounds like if you don't use that "② attribute" which is `#[r#use]`/`#[import]` you can't use any "① tags from other modules" even via full path.

zjp-CN commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,485 @@
+Syntax of a safety tag is defined as follows:
+
+```text
+SafetyTag -> `#` `[` `clippy::safety` `{` Tags `}` `]`
+
+Tags -> Tag (`;` Tag)*
+
+Tag -> ID (`,` ID)* (`:` LiteralString)?
+
+ID -> SingleIdent | SimplePath
````
Good idea! The proposed syntax is inspired by structured safety comments. [Previous discussion.](https://internals.rust-lang.org/t/pre-rfc-safety-property-system/23252/22)

zjp-CN committed:
Update text/0000-safety-tags.md

Co-authored-by: kennytm <kennytm@gmail.com>

zjp-CN committed:
safety-tags: Should Tags Take Arguments?

zjp-CN committed:
safety-tags: rephrase motivations

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+     function’s own signature confirms these tags are listed in its `#![clippy::safety::r#use(...)]`
+     attribute.
+   * Tags are only discharged inside or onto an `unsafe fn`; it's an error to tag a safe function.
+   * If an unsafe call lacks any required tag, emit a diagnostic whose severity (warning or error)
+     is governed by the configured Clippy lint level.
```
I think this is what I updated in motivation: [Granular Unsafe: How Small Is Too Small?](https://github.com/Artisan-Lab/rfcs/blob/80cf23dad9206851906c7b0503a3fcb143227a8c/text/0000-safety-tags.md?plain=1#L72-L79), and safety tags are of [semantic granularity](https://github.com/Artisan-Lab/rfcs/blob/80cf23dad9206851906c7b0503a3fcb143227a8c/text/0000-safety-tags.md?plain=1#L102-L103).

An example of tagging non unsafe block is [propagating safety requirements](https://github.com/Artisan-Lab/rfcs/blob/80cf23dad9206851906c7b0503a3fcb143227a8c/text/0000-safety-tags.md?plain=1#L192-L200).

---

Thank you for reviews and questions. I'll try to catch up them tomorrow due to time difference.

zjp-CN replied:
> procedure should be something like:

Thank you.  I've updated the whole paragraph with your suggestions.

zjp-CN committed:
safety-tags: replace clippy namespace with safety namesapce

Also
* distinguish safety::requires and safety::checked
* replace safety::r#use with safety::import
* only allow uninhabited enum as tag item

zjp-CN committed:
safety-tags: alternative named argument syntax

zjp-CN committed:
safety-tag: Rationale for the Proposed Implementation

zjp-CN committed:
safety-tags: teams support for this RFC

zjp-CN committed:
safety-tags: Why Not Structured Safety Comments?

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,471 @@
+* `#![feature(custom_inner_attributes)]` for `#![clippy::safety {}]` imports
+
+Since the safety-tag mechanism is implemented primarily in Clippy and rust-analyzer, no additional
+support is required from rustc.
```
I've added "[Rationale for the Proposed Implementation](https://github.com/Artisan-Lab/rfcs/blob/2130878acbbc38a481c1aa18734f364d2076ad93/text/0000-safety-tags.md#rationale-for-the-proposed-implementation)" section and extract comparison with structured safety comments to section called "[Why Not Structured Safety Comments?](https://github.com/Artisan-Lab/rfcs/blob/2130878acbbc38a481c1aa18734f364d2076ad93/text/0000-safety-tags.md#why-not-structured-safety-comments)".

zjp-CN committed:
safety-tags: Encapsulate Declaration with define_safety_tag!

zjp-CN committed:
safety-tags: minor clarification

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+Tags constitute a public API; therefore, any alteration to their declaration or definition must be
+evaluated against [Semantic Versioning][semver].
+* Adding a tag declaration or definition is a **minor** change.
+* Removing a tag declaration or definition is a **major** change.
```
Tag declaration and definition are different things:

```rust
#[safety::declare_tag(...)]  # declare a tag item (currently as an uninhabited enum)
enum Declaration {}

#[safety::requires(...)]     # define tags on an unsafe functions
unsafe fn attr_above_is_a_tag_definition() {}
```

For renaming a tag, I don't think it's a problem. As you observed, renaming is a two-step operation as removal and addition. The first step has already incurred a major change. As for changing the meaning of a present tag, we should mark tag item as deprecated or remove it, then define a new tag for the meaning.

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+Tags live in their own [type namespace] carry item-level [scopes] and obey [visibility] rules,
+keeping the system modular and collision-free. Since they are never referenced directly as real
+items, we propose importing them uses a dedicated syntax: inner-styled or outer-styled 
+`clippy::safety::r#use` tool attribute on modules:
```
It's not because "preventing tag usage as normal items". All tag items are used by tag operation `requires` and `checked`, we have to say there is a tag named ValidPtr, and it must be in scope. 

Conditional use statements don't convey the semantics that safety attributes are using tag items, instead it says tag items are used in  logics of user code. Surely people can use tag items if they want. Beside, import tag items via `use` will trigger unused items lints, while they are used in tags.

In this sense, importing tags are a thing from `safety` namespace, so `#![safety::import]` is the right way to express it.

Also, the paragraph is outdated, especially on tag items being uninhabited enums only.

I'm closing this conversation. Feel free to comment on updated contents.

zjp-CN committed:
safety-tags: summarize safety attributes with 💡

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+   * Resolve the callee, collect its declared tags, then walk outward from the call site until the
+     function’s own signature confirms these tags are listed in its `#![clippy::safety::r#use(...)]`
+     attribute.
+   * Tags are only discharged inside or onto an `unsafe fn`; it's an error to tag a safe function.
```
> If the RFC really means that tags are discharged on unsafe calls

Yeah, that's right. I'll update the doc and let's focus the discussion in [this](https://github.com/rust-lang/rfcs/pull/3842#discussion_r2248289374) conversation. Closing this.

zjp-CN committed:
safety-tags: explain how safety::checked is validated on unsafe operation

Other major clarification:
* safety::checked is mandatory in unsafe delegation example
* discussion on precedence of grouped tags

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+3. Verify that every unsafe call carries the required safety tags:
+   * Resolve the callee, collect its declared tags, then walk outward from the call site until the
+     function’s own signature confirms these tags are listed in its `#![clippy::safety::r#use(...)]`
+     attribute.
```
Yeah, it's a typo and should be in `safety::requires`. I've updated the whole paragraph.

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,485 @@
+The process of verifying whether a tag is present is referred to as tag discharge.
+
+Note that it's allowed to discharge tags of unsafe callees onto the unsafe caller for unsafe
+delegation or propogation:
```
Thanks for the advice. I've updated the doc and added a "[Discharge Tags](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#discharge-tags)" section.

zjp-CN committed:
safety-tags: minor adjustment

zjp-CN committed:
safety-tags: semantic granularity with future entity-reference

also add https://github.com/rust-lang/rfcs/pull/3768 safety blocks
to alternative proposals

zjp-CN committed:
safety-tags: minor fixes

ia0 commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,799 @@
+    pub enum ValidPtr() {}
+}
+
+#[safety::requires { ValidPtr }] // 💡 define tags on unsafe functions
```
I don't think grouped tags should be a big issue. Reasons are only for humans, they can figure out to what they apply. Tools simply ignore the reasons, so they just see a list of tags.

Humans can use different conventions:
- `checked { Foo, Bar = "reason for Foo and Bar" }`
- `checked { Foo = "see Bar", Bar = "reason for Foo and Bar" }`

It's more important to get the overall logic right. Where are tags defined? In some empty enums or on each unsafe function?

burdges commented:
@zjp-CN > Do you indicate tags work bad on nested unsafe calls?

I've no opinion on this RFC, or the previous one, or when one should create some static analysis plugin interface.

I'm only pointing out that an unsafe counter achieves lots, without increasing the langauge complexity much, so imho most developers would use an unsafe counter, not a claim we can make for more complex schemes.  That doesn't say those schemes do not have a place.

kennytm commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,788 @@
+
+4. Validate `#[safety::checked]` on HIR nodes whose `ExprKind` is one of
+   - **direct unsafe nodes**: `Call`, `MethodCall` that invoke an unsafe function/method, or
+   - **indirect unsafe nodes**: `Block` (unsafe), `Let`, `Assign`, `AssignOp`.
```
> There's also unsafe attributes

since this RFC only concerns about expressions I think discharging unsafety on items would be out-of-scope but a possible future extension:

* `unsafe impl` an `unsafe trait`
* using an `#[unsafe(...)]` attribute
* declaring a `safe` item inside `unsafe extern`
* using `global_asm!`

> The current RFC as designed is indeed specific to library UB (actually only unsafe functions, not even unsafe traits) and doesn't address language UB (the rest of the unsafe super-powers).

Yeah but the tag name `ValidPtr` sounds very language-UB-y 🙃.

ia0 replied:
> since this RFC only concerns about expressions I think discharging unsafety on items would be out-of-scope

Ah good point. I didn't notice you focused on expressions. So indeed, besides focusing on expressions, this RFC also focuses on unsafe function calls specifically.

> Yeah but the tag name `ValidPtr` sounds very language-UB-y 🙃.

Yes, library UB can simply bubble-up the requirements from language UB to their API if they can't contain it. This is what functions like `ptr::read()` and `ptr::write()` have to do. So it makes sense to name those function call requirements as the requirements of the unsafe superpower the function uses internally.

zjp-CN replied:
I've added ["Tagging More Unsafe Operations"](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#tagging-more-unsafe-operations) section to extend tagging on unsafe operations. I'm closing this conversation. But feel free to comment on the section.

ia0 commented:
> `checked` can be attached to an unsafe block as long as the block containing single unsafe call (no restriction on how many safe calls are in it).

Right. What I meant was to also permit this when there are multiple unsafe calls within the unsafe block (or expression) where the `checked` clause is present. The semantics would be that the `checked` clause needs to contain the union (with or without duplicates, TBD) of the unsafe call tags. That's not "clean", but it gives some amount of freedom to users that could help in such situations where multiple unsafe calls are chained. This is a trade-off essentially.

zjp-CN committed:
remove safety::import; named arguments in safety::{requires,checked}

zjp-CN committed:
add "Tagging More Unsafe Operations"

zjp-CN committed:
minor fixes

zjp-CN committed:
force description in tag definitions; fix `:` by `=`

zjp-CN commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,672 @@
+corresponding tags on the function declaration and mark each one off at the call site.
+
+```rust
+#[safety::requires { ValidPtr, Aligned, Initialized }] // defsite or definition
````
Good point! I've updated the proposal.

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,672 @@
+
+    #[safety::checked {
+      Aligned: "alignment of ptr has be adjusted";
+      ValidPtr, Initialized: "delegated to the caller"
```
Yes. I've corrected them.

zjp-CN committed:
add @deprecated disgnostics; major change to add or remove tags

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,672 @@
+against [Semantic Versioning][semver].
+* Adding a tag definition is a **minor** change.
+* Removing a tag definition is a **major** change. Renaming a tag definition is a two-step
+  operation of removal and addition, bringing a major change due to removal. 
```
`@deprecated` is for better diagnostics.

Your proposal:

```rust
#[safety::requires {
  Tag1 = "...",
  Tag2 = "description",
}]
unsafe fn deprecate_a_tag() {}

// error: Tag2 is not discharged.
// warning: Tag0 is not defined on this function.
#[safety::requires { Tag0, Tag1 }]
unsafe { deprecate_a_tag() }
```

Problems with this design:
* the error says nothing about the relation between Tag0 and Tag2
  * the user has to check out release note, or discussion, or PR to know what happened
* when many tags are deprecated at the same time, the user needs more efforts to know why

 ```rust
#[safety::requires {
  Tag2 = "description",
}]
unsafe fn deprecate_tags() {}

// error: Tag2 is not discharged.
// warning: Tag0 is not defined on this function.
// warning: Tag1 is not defined on this function.
#[safety::requires { Tag0, Tag1 }]
unsafe { deprecate_a_tag() }
```

---

`@deprecated` proposal in this RFC:

```rust
#[safety::requires {
  Tag1 = "...",
  Tag2 = "description",
  Tag0 = "@deprecated Variant must be upheld with variant B now, \
         so this Tag will be removd in the future. Use Tag2 instead.",
}]
unsafe fn deprecate_a_tag() {}

// warning: Tag0 is deprecated: Variant must be upheld with variant B now,
// so this Tag will be removd in the future. Use Tag2 instead.
// error: Tag2 is not discharged.
#[safety::requires { Tag0, Tag1 }]
unsafe { deprecate_a_tag() }
```

```rust
#[safety::requires {
  Tag2 = "description",
  Tag0 = "@deprecated Variant must be upheld with variant B now, \
         so this Tag will be removd in the future. Use Tag2 instead.",
  Tag1 = "@deprecated No need to uphold this invariant. Will be removed soon.",
}]
unsafe fn deprecate_tags() {}

// warning: Tag0 is deprecated: Variant must be upheld with variant B now,
// so this Tag will be removd in the future. Use Tag2 instead.
// warning: Tag1 is deprecated: No need to uphold this invariant. Will be removed soon.
// error: Tag2 is not discharged.
#[safety::requires { Tag0, Tag1 }]
unsafe { deprecate_tags() }
```

Ofc you may argue that if invariants change drastically like this, we should mark `#[deprecated]` on `deprecate_tags`, and point to another unsafe function.

zjp-CN committed:
minor fix

Co-authored-by: kennytm <kennytm@gmail.com>

alice-i-cecile commented:
This flavor of thing would be nice to introduce more structure (and the possibility for automated tooling) into complex unsafe code bases like `bevy_ecs`, which I help maintain.

alice-i-cecile commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,662 @@
+[#135805]: https://github.com/rust-lang/rust/pull/135805
+[#135009]: https://github.com/rust-lang/rust/pull/135009
+
+Even if the safety documentation is complete, two problems remain:
```
There is another important benefit that should be called out here: what happens if the safety invariants change? Updating callsites (including in downstream consumer code) is very important and extremely easy to overlook.

zjp-CN replied:
Yeah. That's what I wrote in [Safety Invariants Have No Semver](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#safety-invariants-have-no-semver).

zjp-CN committed:
remove `@deprecated`; swap the order on Versioned invariants and Semantic granularity

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,647 @@
+   * To enable truly semantic checking, we envision an [entity-reference] system that meticulously
+     traces every unsafe related operation that could break an invariant in source code.
+
+4. **Lightweight checking**. Clippy only matches tag paths. No heavyweight formal proofs, keeping
```
As there's no tag items, the scope of a tag is limited to the unsafe function, thus two unsafe functions can have identical tag name but without ambiguity, right? Also, there's never name collision with tags from downstream crates!

On another words, we must tag on single unsafe operation instead unsafe blocks containing multiple unsafe operations to prevent ambiguity or collision. Good!

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,647 @@
+Tags constitute a public API; therefore, any alteration to their definition must be evaluated
+against [Semantic Versioning][semver].
+* Adding a tag definition is a **major** change, because new tag is missing. 
+* Removing a tag definition is a **major** change, because the tag doesn't exist.
```
> I don't see anything suggesting that extra discharged tags are bad.

It's bad when discharging the extra tag that happens to be added to the unsafe operation, with possiblity of failure to manually verify if the tag/invariant is held when no one noticed the  warining of absent definition.

kennytm replied:
> It's bad when discharging the extra tag that happens to be added to the unsafe operation after a while, with possiblity of failure to manually verify if the tag/invariant is upheld when no one noticed the warning of absent definition.

So consider the scenario, which the callee crate initially put `#[safety::requires(A = "condition 1")]` at v1.0, removed the tag at v1.1, and put the *same tag* back *with a different requirement* `#[safety::requires(A = "condition 2")]` at v2.0.

IMO the fault is not in the caller crate still having `#[safety::checked(A = "certification 1")]` using v1.1, because if the caller upgraded its dependency from v1.0 to v2.0 there will be no errors at all despite the requirement about `A` is totally changed.

So what's more important is ia0's suggest that "*Changing the definition of a tag in a way that requires more, is a major change*". I'd even say that this should be **prohibited** across all major versions, and if the callee crate author wants to reintroduce the same tag on the same item it must reuse the same description, or choose a different tag name.

zjp-CN replied:
Thank you to both of you! I've [updated](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#safety-tags-are-a-part-of-an-unsafe-function) as per your suggestions.

ia0 commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,647 @@
+each tag name, its optional description, and possible deprecated state below the tagged function or
+other unsafe item.
+
+## Generate Safety Docs from Tags
```
Good point! And ideally, it could even add some short notice that those are safety tags, for example with a link to some HOWTO about calling an unsafe function with safety tags.

zjp-CN replied:
Thank you! I've moved "[Auto Generate Safety Docs from Tags](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#auto-generate-safety-docs-from-tags)" into our proposal in https://github.com/rust-lang/rfcs/pull/3842/commits/21e54312487def28899ee05fb0e046ce4c246c2b .

zjp-CN committed:
minor fix

Co-authored-by: Julien Cretin <github@ia0.eu>

zjp-CN committed:
clarify semver on definition change

zjp-CN committed:
Auto Generate Safety Docs from Tags; multiple attrs on the same node

zjp-CN committed:
minor fixes and clarification on semver

zjp-CN committed:
minor fixes: discharge of undefined tags warns by default

danjl1100 reviewed (commented):
I really like the direction this is moving with `requires` and `checked` since the initial version.

zjp-CN committed:
minor fix

ia0 commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,692 @@
+
+Tags constitute a public API; therefore, any alteration to their definition must be evaluated
+against [Semantic Versioning][semver].
+* Adding a tag definition is a **major** change, because new tag is missing. 
```
That's an interesting point and I'm curious how this is handled for regular API changes, namely:

```diff
 /// Returns the quotient of 2 numbers.
-pub fn my_div(x: u32, y: u32) -> u32;
+pub fn my_div(x: u32, y: NonZero<u32>) -> u32;
```

Is this an breaking change?

If yes, how does it differ from adding a safety pre-condition?

hxuhack commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,692 @@
+[more unsafe operations]: #tagging-more-unsafe-ops
+[entity-reference]: #entity-reference
+
+## `#[safety]` Tool Attribute and Namespace
+
+Safety tags have no effect on compilation; they merely document safety invariants. Therefore, we
+propose that `#[safety]` be implicitly registered for every crate.
+
+```rust
+#![feature(register_tool)]
+#![register_tool(safety)]
+```
````
Through the `safety` namespace, we aim to tame `unsafe` code and address Rust’s last-mile problem in memory safety. The `safety` namespace is not only used by Clippy and rust-analyzer, but also serves as a source of safety guidance for developers (see _[Auto Generate Safety Docs from Tags](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#auto-generate-safety-docs-from-tags)_). By contrast, the [diagnostic](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-diagnostic-tool-attribute-namespace) namespace is primarily intended for customizing debugging messages. Mixing the two may not be advisable.

Furthermore, we may extend the `safety::requires` tags with parameters in the future, enabling them to act as formal safety specifications, similar to contracts. This feature is outside the scope of the current RFC (see _[Safety Properties with Arguments for Verification](https://github.com/Artisan-Lab/tag-std/blob/main/pre-RFC.md#safety-properties-with-arguments-for-verification)_ for more details).

zjp-CN committed:
review: request Lang team to reserve `safety` namespace

Co-authored-by: Mads Marquart <mads@marquart.dk>

zjp-CN committed:
syntax: only accepy `()` delemiter around tags

cc https://github.com/rust-lang/rfcs/pull/3842#discussion_r2293184928

zjp-CN committed:
`checked` is lenient in diagnostics by default; upgrade in cargo check

zjp-CN committed:
tag styles: visually group tags; naming convention

zjp-CN committed:
rephrase team approvals

zjp-CN committed:
entity-reference attribute candidates due to `ref` being a keyword

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,672 @@
+      and its nearest ancestor that is an *indirect unsafe node*.
+   2. If the callee is annotated with safety tags, require that **either** the call itself **or**
+      its recorded ancestor carries `#[safety::checked]`.
+   3. Any node that carries `#[safety::checked]` must contain **exactly one** unsafe call/method;
+      otherwise emit a diagnostic. *(We intentionally stop at this simple rule; splitting complex
+      unsafe expressions into separate annotated nodes is considered good style.)*
```
Supporting such tagging means that tags in `checked` are applied to multiple unsafe operations. This results in obligations being discharged on a large scale, including nested and chained calls, as well as the same calls for different values:

```rust
unsafe {
    #[safety::checked { ValidPtr }] // `ptr` and temporary `&ptr.read() as *const ()` are valid to read
    ptr.read()
       .read()
};
```

I’d say this could lead to confusion and potential misuse, making it unclear what has been checked unless both pointers are explicitly mentioned in the reason.

I think it's less error-prone to let `safety::checked` handle atomic discharging, and introduce a new attribute named `safety::batch_checked` to support the discharging of shared tags.

```rust
unsafe {
    #[safety::batch_checked {
        ValidPtr = "`ptr` and temporary `&ptr.read() as *const ()` are valid to read"
    }]
    ptr.read()
       .read()
};

    #[safety::batch_checked {
        Aligned = "arrays are properly aligned",
        ValidForReads = "the arrays are owned by this function, and contain the copy type f32",
    }]
    unsafe {
        float32x4x4_t(
            vld1q_f32(a.as_ptr()),
            vld1q_f32(b.as_ptr()),
            vld1q_f32(c.as_ptr()),
            vld1q_f32(d.as_ptr()),
        )
    }
```

zjp-CN replied:
> I'd argue the same is true of `unsafe`, and we have the [`multiple_unsafe_ops_per_block`](https://rust-lang.github.io/rust-clippy/master/index.html#multiple_unsafe_ops_per_block) lint to deal with that. I'd expect to have a similar lint when using `#[safety::checked]` to check multiple unsafe operations.

That would echo the motivation ["Granular Unsafe: How Small Is Too Small?"](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#granular-unsafe-how-small-is-too-small).

Our principle is **semantic granularity**, so I'd just propose a new attribute `safety::batch_checked` to express shared mode of discharging, instead of a new high level linter for safety tags.

Proposed as follows

```rust
#[safety::checked(...)] // ❌ hard error
unsafe { char::from_u32_unchecked(*ptr.cast::<u32>()) }

#[safety::batch_checked(...)] // ✅
unsafe { char::from_u32_unchecked(*ptr.cast::<u32>()) }

#[safety::checked(...)] // ✅
let int_value = unsafe { *ptr.cast::<u32>() };
#[safety::checked(...)] // ✅
unsafe { char::from_u32_unchecked(int_value) }
```

instead of

```rust
#[safety::checked(...)] // 🤔 warning of shared_tags_checked
unsafe { char::from_u32_unchecked(*ptr.cast::<u32>()) }

#[allow(clippy::shared_tags_checked)] // 🤔
#[safety::checked(...)] // ✅
unsafe { char::from_u32_unchecked(*ptr.cast::<u32>()) }
```

zjp-CN committed:
tag: `snake_case` naming convention and render to `Snake case`

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,692 @@
+#[safety::requires { // 💡 define safety tags on an unsafe function
+    ValidPtr = "src must be [valid](https://doc.rust-lang.org/std/ptr/index.html#safety) for reads",
+    Aligned = "src must be properly aligned, even if T has size 0",
+    Initialized = "src must point to a properly initialized value of type T"
```
I've updated RFC at https://github.com/rust-lang/rfcs/pull/3842/commits/23f89428db890f7f911f2cfbd095571ae6cfa6ef to state that `snake_case` naming convention for tags, and how rustdoc should render them.

zjp-CN committed:
future: `#[safety::batch_checked]` Shares Tag Discharging

Alexendoo commented:
https://github.com/rust-lang/rust-clippy/pull/11600 was a related effort on the clippy side that stalled on inactivity, though not specific to `unsafe`

zjp-CN committed:
chore: typos

zjp-CN committed:
alternativs: compared to Clippy lint `danger_not_accepted`

Escarcega1989 commented on text/0000-safety-tags.md:
```diff
@@ -565,16 +563,16 @@ We could allow *any* arguments in tag usage without validation. Tag arguments wo
   }
-}]
+)]
 unsafe fn foo<T>(ptr: *const T) -> T { ... }
```
unsafe fn foo<T>(ptr: *const T) -> T { ... }

Escarcega1989 commented on text/0000-safety-tags.md:
```diff
@@ -565,16 +563,16 @@ We could allow *any* arguments in tag usage without validation. Tag arguments wo
 
-#[safety::checked { ValidPtr(p) }] // p will not be type-checked
+#[safety::checked(ValidPtr(p))] // p will not be type-checked
 unsafe { bar(p) }
```
"Unsafe { bar}"

joshlf commented:
Heads up that the [unsafe fields RFC](https://github.com/rust-lang/rfcs/pull/3458) (which is referenced in this RFC) has now been merged.

cc @jswrenn

zjp-CN commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,808 @@
+  an undefined tag just emits a warning-by-default diagnostic.
+* Renaming a tag definition is a **major** change, because it's the result of removal and addition.
+* Changing the definition of a tag in an *equivalent* or in a way that *requires less* (the old tag
+  implies the new tag), is a **minor** change.
```
Good catch!

Weakening the description of `valid_x` and bumping `bar` from 1.0.0 to 1.1.0 makes sense, because any `x` that was required to be "**positive** + **even**" in 1.0.0 still satisfies the safety constraint "**even**" in 1.1.0. (Also see the [previous discussion](https://github.com/rust-lang/rfcs/pull/3842/changes/BASE..5dffa30b514f27e00dc362e1a1854cfc65f8b073#r2256131985   ) on why removing a tag definition is a minor change: in this case, `valid_x` can be treated as two tags `positive_x + even_x` in 1.0.0, and becomes just `even_x` in 1.1.0.)

For soundness, here is my thought: 
If `x` is negative in `bar` 1.0.0, it breaks the safety requirements and can cause safety bugs. However, the same negative `x` will never break the requirement and is always safe to use in `bar` 1.1.0.
But is it safe for `foo` to pass a negative `x` to both `bar` 1.1.0 and 1.0.0? I think the answer is yes, because `bar` chooses to reuse the tag name and remove the requirement  `positive_x` when `bar` doesn't actually rely on the `positive_x` requirement for correctness.
In other words, if `bar` relied on the fact that `positive_x` causes safety bugs, it couldn't simply modify the description without renaming the tag! That is to say: relaxing a tag description will never cause unsoundness **if and only if** the removed requirement is nowhere relied upon.

zjp-CN replied:
> I just realized that this assumes people use the correct version requirement. I'm not sure this is a realistic assumption.

For this concern, I'd suggest renaming the tag with a new description whenever you're unsure if anything relies on the relaxed behavior: this keeps semver working correctly.

zjp-CN replied:
Thanks for the concise wording! I've [updated](https://github.com/rust-lang/rfcs/pull/3842/commits/a20f5105159fa86a73c179163bb69c768939363a) it.

zjp-CN committed:
semver: rename tag when relaxed requirement affects old versions

cc https://github.com/rust-lang/rfcs/pull/3842#discussion_r2804411205

zjp-CN commented:
I just noticed a case where safety invariants in libcore changed in a semver-breaking manner as introduced in https://github.com/rust-lang/rust/pull/141260 and stablized since 1.90: for read_volatile/write_volatile on [latest stable](https://doc.rust-lang.org/stable/std/ptr/fn.read_volatile.html) vs [an old version such as 1.88](https://doc.rust-lang.org/1.88.0/std/ptr/fn.read_volatile.html), the safety doc section differs

```diff
-src must be valid for reads.
+src must be either valid for reads, or it must point to memory outside of all Rust allocations 
+and reading from that memory must:
+* not trap, and
+* not cause any memory inside a Rust allocation to be modified.
```

Since maintaining invariant semver is a core motivation for safety tags, relying solely on [release notes](https://doc.rust-lang.org/stable/releases.html#version-1900-2025-09-18) is insufficient; automated tool checks are essential to track such changes effectively.

```diff
-#[safety::requires(valid_ptr(src))]
+#[safety::requires(any(valid_ptr(src), not_rust_allocation(src)))]
```

ThierryBerger commented on text/0000-safety-tags.md:
````diff
@@ -0,0 +1,809 @@
+
+The process of verifying whether a tag is checked is referred to as tag discharge.
+
+Now consider forwarding invariants of unsafe callees onto the unsafe caller for unsafe delegation or
+propogation:
+
+```rust
+#[safety::requires(valid_ptr = "...", aligned = "...", initialized = "...")]
+unsafe fn propogation<T>(ptr: *const T) -> T {
+    #[safety::checked(valid_ptr, aligned, initialized)]
+    unsafe { read(ptr) }
+}
+```
````
`propogate` should be `propagate` (**a**)

```suggestion
Now consider forwarding invariants of unsafe callees onto the unsafe caller for unsafe delegation or
propagation:

```rust
#[safety::requires(valid_ptr = "...", aligned = "...", initialized = "...")]
unsafe fn propagation<T>(ptr: *const T) -> T {
    #[safety::checked(valid_ptr, aligned, initialized)]
    unsafe { read(ptr) }
}
```
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3842/b6d0de380d85d96eba68fb6accf34d6c5fb3bd7f..a20f5105159fa86a73c179163bb69c768939363a)*

DanielEScherzer commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,809 @@
+than the code itself, which makes them error-prone and increases the risk that reviewers may
+overlook inaccuracies or missing safety requirements.
+
+Recent Rust issues [#134496],[#135805], and [#135009] illustrate the problem: several libstd APIs
```
```suggestion
Recent Rust issues [#134496], [#135805], and [#135009] illustrate the problem: several libstd APIs
```

missing a space here

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3842/b6d0de380d85d96eba68fb6accf34d6c5fb3bd7f..a20f5105159fa86a73c179163bb69c768939363a)*

DanielEScherzer commented on text/0000-safety-tags.md:
```diff
@@ -0,0 +1,809 @@
+
+## Formal Contracts, Casual Burden
+
+[Contracts][contracts] excel at enforcing safety invariants rigorously, but they demand the
```
the contracts link no longer works and returns a 404, the goal was not accepted

maybe link to the markdown version at https://github.com/rust-lang/rust-project-goals/blob/main/src/2024h2/Contracts-and-invariants.md ?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3842/b6d0de380d85d96eba68fb6accf34d6c5fb3bd7f..a20f5105159fa86a73c179163bb69c768939363a)*
